File: class02/python/TCC0201-GeradorDePalavras.py
#!/usr/bin/env python3
"""TCC0201-GeradorDePalavras.py | Aula 02
Gerador de palavras para a gramática dos números naturais:
  (1) N -> D
  (2) N -> DN
  (3) D -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9

Cada passo de derivacao e um par (variavel, indice_da_producao).
Para D, o indice_da_producao e o proprio digito (0..9).
Para N, indice 0 = producao (1) "D", indice 1 = producao (2) "DN".
"""

producoes = {
    "N": [["D"], ["D", "N"]],
    "D": [[c] for c in "0123456789"],
}


def deriva(forma, passos):
    """Aplica uma sequencia de (variavel, indice_da_producao) e imprime cada passo."""
    forma = list(forma)
    print("".join(forma))
    for var, regra in passos:
        i = forma.index(var)
        forma = forma[:i] + list(producoes[var][regra]) + forma[i + 1:]
        print("=>", "".join(forma))
    return "".join(forma)


def gerar_para_numero(numero: str):
    """Gera automaticamente a sequencia de passos que deriva a string `numero`."""
    passos = []
    for i, digito in enumerate(numero):
        ultimo = (i == len(numero) - 1)
        passos.append(("N", 0 if ultimo else 1))
        passos.append(("D", int(digito)))
    return passos


if __name__ == "__main__":
    print("=== Derivação de 243 (passos manuais) ===")
    resultado = deriva(["N"], [("N", 1), ("D", 2), ("N", 1), ("D", 4), ("N", 0), ("D", 3)])
    assert resultado == "243"

    print()
    print("=== Derivação automática de 77 ===")
    resultado77 = deriva(["N"], gerar_para_numero("77"))
    assert resultado77 == "77"

    print()
    print("=== Derivação automática de 5890 ===")
    resultado5890 = deriva(["N"], gerar_para_numero("5890"))
    assert resultado5890 == "5890"

File: class02/java/TCC0201-GeradorDePalavras.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * TCC0201-GeradorDePalavras.java | Aula 02
 * Gerador de palavras para a gramática dos números naturais:
 *   (1) N -> D
 *   (2) N -> DN
 *   (3) D -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
 *
 * Mesma lógica e mesma saída do TCC0201-GeradorDePalavras.py.
 */
public class GeradorDePalavras {

    static Map<String, List<List<String>>> producoes = new LinkedHashMap<>();
    static {
        producoes.put("N", List.of(List.of("D"), List.of("D", "N")));
        List<List<String>> digitos = new ArrayList<>();
        for (char c = '0'; c <= '9'; c++) {
            digitos.add(List.of(String.valueOf(c)));
        }
        producoes.put("D", digitos);
    }

    record Passo(String variavel, int regra) {}

    static String deriva(List<String> formaInicial, List<Passo> passos) {
        List<String> forma = new ArrayList<>(formaInicial);
        System.out.println(String.join("", forma));
        for (Passo p : passos) {
            int i = forma.indexOf(p.variavel());
            List<String> substituicao = producoes.get(p.variavel()).get(p.regra());
            forma.remove(i);
            forma.addAll(i, substituicao);
            System.out.println("=> " + String.join("", forma));
        }
        return String.join("", forma);
    }

    static List<Passo> gerarParaNumero(String numero) {
        List<Passo> passos = new ArrayList<>();
        for (int i = 0; i < numero.length(); i++) {
            boolean ultimo = (i == numero.length() - 1);
            passos.add(new Passo("N", ultimo ? 0 : 1));
            passos.add(new Passo("D", numero.charAt(i) - '0'));
        }
        return passos;
    }

    public static void main(String[] args) {
        System.out.println("=== Derivação de 243 (passos manuais) ===");
        String resultado = deriva(List.of("N"), List.of(
                new Passo("N", 1), new Passo("D", 2),
                new Passo("N", 1), new Passo("D", 4),
                new Passo("N", 0), new Passo("D", 3)));
        assert resultado.equals("243");

        System.out.println();
        System.out.println("=== Derivação automática de 77 ===");
        String resultado77 = deriva(List.of("N"), gerarParaNumero("77"));
        assert resultado77.equals("77");

        System.out.println();
        System.out.println("=== Derivação automática de 5890 ===");
        String resultado5890 = deriva(List.of("N"), gerarParaNumero("5890"));
        assert resultado5890.equals("5890");
    }
}
